client-go/tools/portforward: Add ErrorHandler field

brianpursley opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Some server processes will send RST while a connection is being closed, resulting in the handled connection terminating with "connection reset by peer" on the error stream.

This commit adds an ErrorHandler field to PortForwarder, which, if set, provides a way to handle errors before PortForwarder's default error handling occurs. If the function is nil, or if the function returns false, then PortForwarder's default error handling will occur. If the function returns true, then PortForwarder will not perform any additional error handling.

This commit also updates kubectl's portforward command implementation to handle "connection reset by peer" by only logging it and returning true so that it will be treated as a non-fatal error and the port forwarding session is not terminated.

#### Which issue(s) this PR fixes:
Fixes #111825 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Port forwarding changed to handle connection reset by peer without terminating the port forwarding session
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

brianpursley commented:
/assign @liggitt

Jeansen commented:
Am I understanding this right, that you check for an explicit string of " "connection reset by peer", @brianpursley If that string changes, the PR will no longer be valid. Wouldn't it be better to make it more generic, like checking for this RST Package, if possible?

brianpursley commented:
> Am I understanding this right, that you check for an explicit string of " "connection reset by peer", @brianpursley If that string changes, the PR will no longer be valid. Wouldn't it be better to make it more generic, like checking for this RST Package, if possible?

Yeah, I get what you're saying.  In `handleConnection()`, it is operating at a hugher level, so it doesn't know anything about individual packets, it just has go errors (ie. read failed, write failed) and the error stream.

FWIW, I tried to do something similar to this in api machinery:
https://github.com/kubernetes/kubernetes/blob/e1069c64956a43f628d8ae2fcd9107959747c1c6/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L97

Perhaps the logic of non-fatal or transient errors could be centralized in API machinery, so it is only in one place, but for starters I am curious to get some feedback about the approach.  But I am open to refactoring to avoid duplication.

aojea commented:
> Perhaps the logic of non-fatal or transient errors could be centralized in API machinery, so it is only in one place, but for starters I am curious to get some feedback about the approach. But I am open to refactoring to avoid duplication.

those are http errors, that we should not generalize for all TCP protocols

aojea commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -355,7 +355,7 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 		case err != nil:
 			errorChan <- fmt.Errorf("error reading from error stream for port %d -> %d: %v", port.Local, port.Remote, err)
-		case len(message) > 0:
+		case len(message) > 0 && !strings.Contains(string(message), "connection reset by peer"):
```
your current connection will be closed anyway by the operating system, you'll need to create  a new one

aojea replied:
does this not exit anyway?
I can see that close(errorChan) is always called

.. but `waitForConnection` keeps running :thinking:

brianpursley replied:
Yes the errChan is closed and handleConnection() exits, but since <-errorChan returns nil then it will not close streamConn (see line 405), so waitForConnection will be able to continue listening for and accepting new connections.

brianpursley commented:
> > Perhaps the logic of non-fatal or transient errors could be centralized in API machinery, so it is only in one place, but for starters I am curious to get some feedback about the approach. But I am open to refactoring to avoid duplication.
> 
> those are http errors, that we should not generalize for all TCP protocols

Ah you're right.  I didn't notice that at first.  I agree they shouldn't be mixed.

liggitt commented:
a few comments:
1. stringly typed error detection is pretty fragile... if this message changes in the future, this will break
2. there are other cases handled in staging/src/k8s.io/apimachinery/pkg/util/net/util.go and staging/src/k8s.io/apimachinery/pkg/util/net/http.go ... do any of them apply here?
3. what are the consequences if this heuristic returns a false positive if this string is observed in a response that does not represent a peer connection reset between kubectl and apiserver? is it possible to get into an infinite retry loop?

brianpursley commented:
> a few comments:
> 
> 1. stringly typed error detection is pretty fragile... if this message changes in the future, this will break
> 2. there are other cases handled in staging/src/k8s.io/apimachinery/pkg/util/net/util.go and staging/src/k8s.io/apimachinery/pkg/util/net/http.go ... do any of them apply here?
> 3. what are the consequences if this heuristic returns a false positive if this string is observed in a response that does not represent a peer connection reset between kubectl and apiserver? is it possible to get into an infinite retry loop?

Thank you @liggitt 

You're comments motivated me to look deeper and you're right, there is a way to check for and handle ECONNRESET and not the specific error text. Please take another look and see what you think.

Also, there was some existing code that used stringly typed error detection ("use of closed network connection") in three places. Since I was touching that part of the code, I refactored those into a single function at least. Do you think I should change this to compare `err == net.ErrClosed` instead, or leave as a string?  The comments here seem to indicate that it should be OK to compare to `net.ErrClosed` instead of checking the string:
https://cs.opensource.google/go/go/+/master:src/net/error_test.go;drc=85b5f86584686677c554b2538a7edee96d684aea;l=502-514
```
// Because historically we have not exported the error that we
// return for an operation on a closed network connection,
// there are programs that test for the exact error string.
// Verify that string here so that we don't break those
// programs unexpectedly. See issues #4373 and #19252.
want := "use of closed network connection"
if !isShutdown && !strings.Contains(nestedErr.Error(), want) {
    return fmt.Errorf("error string %q does not contain expected string %q", nestedErr, want)
}

if !isShutdown && !errors.Is(nestedErr, ErrClosed) {
    return fmt.Errorf("errors.Is(%v, errClosed) returns false, want true", nestedErr)
}
```

aojea commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -404,10 +408,20 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 	}
 }
 
+func isUseOfClosedNetworkConnection(err error) bool {
+	return strings.Contains(strings.ToLower(err.Error()), "use of closed network connection")
+}
```
this has to be something legacy, you have this error if someone is closing the listenet , with `listener.Close()`

aojea replied:
I see, there is no graceful shutdown, that is why it fails and silent those errors

brianpursley replied:
Yeah, I wasn't completely sure of the reason for that check.  I suspect this could be replaced by `errors.Is(err, net2.ErrClose)` but I'm not 100% certain it is the same thing.

brianpursley commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -374,20 +376,22 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 
 		// Copy from the local port to the remote side.
-		if _, err := io.Copy(dataStream, conn); err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
+		if _, err := io.Copy(dataStream, conn); err != nil && err != net.ErrClosed {
```
I was experimenting and forgot to undo that.  My mistake sorry (I'm trying to do too many parallel things today :thinking:)

It should have been `errors.Is(err, net.ErrClosed)` anyway, but I'm not sure if that is a valid check or not, so I was not going to try to do that yet.

fedebongio commented:
/triage accepted

aojea commented:
I was checking this deeper and I think that the problem is not to retry, is that the mix of different scenarios and the organically growing of the tool, I suggest to go back to the blackboard instead of keep growing unless we have clear and reproducible bugs. 

So, portforward is a basic proxy it

1. opens a local listener
2. opens a connection to a remote pod based on an object type passed as a parameter pod/service/deployment, this is encapsulated (uses SPDY or websocket) and tunneled through the apiserver -> kubelet -> CRI -> pod
3. it starts copying from 1 to 2 and viceversa

The local connection 1. is one side that we control, this is straightforward.
The part 2 is the problematic because we don't really know what is going on, we only have the httpstream information but there are at least 3 jumps we are blind, we have to decide what do we want to do and how to handle that side of the connection.

Based on the description of the command, the "proxy" lifecycle should match the destination pod

```
Forward one or more local ports to a pod.

 Use resource type/name such as deployment/mydeployment to select a pod. Resource type defaults to 'pod' if omitted.

 If there are multiple pods matching the criteria, a pod will be selected automatically. The forwarding session ends
when the selected pod terminates, and a rerun of the command is needed to resume forwarding.
```

so it seems legit to keep forwarding meanwhile the httpstream is alive

brianpursley commented:
@aojea, thanks for the insight, see my responses below to a few of your points.

> I was checking this deeper and I think that the problem is not to retry

As mentioned over on the issue (https://github.com/kubernetes/kubernetes/issues/111825#issuecomment-1217963766), but I'll comment here for the discussion, I probably wasn't clear enough, but this is not performing any retry logic.  It only is leaving the listener open (pf.streamConn, which unfortunately is not well named, it is really the listener connection). So the listener will keep running and be able to handle subsequent connections. Otherwise, how it works now, when ECONNRESET occurs, the listener is stopped and kubectl exits, so the user has to repeatedly restart port forwarding when they should not need to.

> I suggest to go back to the blackboard instead of keep growing unless we have clear and reproducible bugs

I agree that port forwarding could use an overhaul, but there are clear and reproducible bugs, such as #111825, where port forwarding terminates unexpectedly when you are using psql to communicate with PostgreSQL in SSL mode, and issue certain commands (`\c` for example).

The root cause seems to be that PostgreSQL doesn't cleanly handle closing a connection and sends RST, resulting in an ECONNRESET error. Now I suppose you could say that PostgreSQL shouldn't do that, but for kubectl port-forward's part, I believe it should not abandon the entire port forwarding session (ie. close pf.streamConn) in this case because the remote is still there.

Also, to be clear, it's not just a PostgreSQL problem.  Another user [reported the same problem with argocd](https://github.com/kubernetes/kubectl/issues/1169#issuecomment-1201415049)

So I'm all for re-doing port forwarding, but I think to solve this annoying problem for the community in the near term, that we should see if we can preserve the port forwarding session when ECONNRESET occurs.

brianpursley commented:
To make things worse, the "workaround" people are using for this is to [disable SSL in psql](https://github.com/kubernetes/kubectl/issues/1169#issuecomment-1165140134) or argocd.  So an additional motivation for this change is to avoid causing people to use an insecure workaround.

brianpursley commented:
Ugh. I did some additional testing and handling ECONNRESET does not solve the problem in all cases.  The original change where I looked explicitly for the string "connection reset by peer" does solve it, but in a less than desirable way because it is comparing strings to determine the error.

I'm going to convert this to WIP for now.  It is going to take some more time to figure out the best approach.

aojea commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -404,10 +405,18 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 
+// isFatalError determines whether a given error should result in the closing of the listener connection
+func isFatalError(err error) bool {
+	return !net2.IsConnectionReset(err) &&
```
the error that comes from `errorChan` it seems that is always wrapped

```go
	errorChan := make(chan error)
	go func() {
		message, err := ioutil.ReadAll(errorStream)
		switch {
		case err != nil:
			errorChan <- fmt.Errorf("error reading from error stream for port %d -> %d: %v", port.Local, port.Remote, err)
		case len(message) > 0:
			errorChan <- fmt.Errorf("an error occurred forwarding %d -> %d: %v", port.Local, port.Remote, string(message))
		}
		close(errorChan)
	}()
```

can this ` !net2.IsConnectionReset(`  happen?

Jeansen commented:
I think it is OK, too. I am not very familiar with the code but one rule above all is to definitely to do a refactoring in little steps; one action at a time, so if things break it is clear what change broke it. So, thanks for your efforts, @brianpursley

brianpursley committed:
client-go/tools/portforward: Add ErrorHandler field

Some server processes will send RST while a connection is
being closed, resulting in the handled connection terminating
with "connection reset by peer" on the error stream.

This commit adds an ErrorHandler field to PortForwarder, which,
if set, provides a way to handle errors before PortForwarder's
default error handling occurs. If the function is nil, or if the
function returns false, then PortForwarder's default error handling
will occur. If the function returns true, then PortForwarder will
not perform any additional error handling.

This commit also updates kubectl's portforward command implementation
to handle "connection reset by peer" by only logging it and returning
true so that it will be treated as a non-fatal error and the port
forwarding session is not terminated.

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/portforward/portforward.go:
```diff
@@ -141,9 +142,22 @@ func (f *defaultPortForwarder) ForwardPorts(method string, url *url.URL, opts Po
+// function logs it and returns true, indicating that the error was handled, and allowing the port forwarder to continue
+// listening for and accepting new connections.
+func portForwardErrorHandler(err error) bool {
+	if err != nil && strings.HasSuffix(strings.ToLower(err.Error()), "connection reset by peer") {
```
lowercase suffix-only match seems super rigid... what layer is producing this error and are we sure there's no layer in between kubectl and that layer that will wrap it or add a suffix?

brianpursley commented:
@liggitt , Thank you for your comments.

I spent some time today tracing through the code to determine the answer to a few questions related to this problem, and hopefully I can shed some more light on what is happening and how it might be handled.  Some of these questions were asked by you, but I paraphrased them a little and added some of my own in the mix as I was investigating.  I hope I captured the spirit of what you were asking at least.

---

**QUESTION: What layer is producing the text (container runtime, kubelet, or apiserver?)**

**RELATED QUESTION: Why can't this error be handled like a normal error? Isn't it wrapped and can't we just use `errors.Is(err, syscall.ECONNRESET)`?**

Short answer: Container Runtime

Long answer: 

I looked at containerd (I realize there can be other container runtimes though), and here is what I found happening...

In containerd, `criService.portforward()` experiences an ECONNRESET error (or any error really) from the pod. It correctly wraps the error and returns it:
https://github.com/containerd/containerd/blob/dcfaa30ba2b43806d6bf71438bd84f9a9c9ffdc4/pkg/cri/server/sandbox_portforward_linux.go#L131-L133

But... 

Higher in call stack, still in containerd, `httpStreamHandler.portforward()` handles err and writes it out as a string on the error stream. This turns it into a string from this point on:
https://github.com/containerd/containerd/blob/3508ddd3dd96796c134196f285d56374ccaeeaeb/pkg/cri/streaming/portforward/httpstream.go#L255-L259

As a result of the above, what comes back on the error stream in client-go's PortForwarder is a pretty big string containing all the wrapped errors that were stringified:
```
 error forwarding port 5432 to pod 187c8a96cb6293d27c62f5bc662cc92b17907f15cd687551ef79015c2541eb55, uid : failed to execute portforward in network namespace "/var/run/netns/cni-09ffdc1b-6536-ef2b-d26e-2ce895699cb0": read tcp4 127.0.0.1:58538->127.0.0.1:5432: read: connection reset by peer
```

Unless there is a way to "re-wrap" an error, this eliminates the possibility of unwrapping the error and using `errors.Is` to detect `ECONNRESET`.

I don't think anything else is intercepting the content of the stream between the container runtime and client-go, so this appears to be coming back to client-go's PortForwarder exactly as it is written.

---

**QUESTION: Where does the string "connection reset by peer" come from?**

This exact string appears to originate from a platform-specific lookup table in Go's x/sys/unix package.
https://cs.opensource.google/go/x/sys/+/master:unix/zerrors_linux_amd64.go;l=759

The text description of ECONNRESET could potentially be different for each platform, but currently it looks to be identical for all platforms.

---

**QUESTION: Is checking the string to see if it has a suffix matching "connection reset by peer" good enough? Will this be too fragile?**

Potential problems:

* x/sys/unix could change the error text
* x/sys/unix could add a new platform that has different text
* containerd could change how it prints the error message in the future.
* Other container runtimes might not even write their error messages out in the same way.

It would be a little less fragile to check for "connection reset by peer" anywhere in the error stream content, instead of just the suffix. This might be the best we can do with having to rely on string matching.

There is definitely a potential problem in that some container runtime might decide to do things differently.

---

**Approaches being considered:**

Assuming string matching is the only way to determine if connection reset by peer occurred, there are two approaches I can think of to handling this:

1. Change client-go's PortForwarder to check for the text "connection reset by peer" anywhere in the error stream content, and if found, don't stop the listener. **This was my original approach to the problem.**

2. Add an optional `ErrorHandler` field to client-go's PortForwarder that allows a client to specify its own error handler, so that it can intercept connection reset by peer. Change kubectl to provide its own ErorrHandler that looks for "connection reset by peer" anywhere in the error message. **This is what this PR is doing currently, but it is only matching the suffix of the error message currently**

Are either of these options acceptable changes to client-go's PortForwarder, and if so, which is the better approach?  Is there a third option I haven't considered?

I like option 1, but understand that it does make it the default behavior for all clients, not just kubectl. I think this seems like it would be OK, but maybe there is a reason a client would not want this behavior, and would want the listener to be stopped.

liggitt commented:
thanks for the super thorough investigation

> Short answer: Container Runtime

Hrmm... coding kubectl to check for a specific error message substring coming from a _specific_ implementation of the container runtime (not from a kubernetes component) seems problematic... is there any precedent for something like this?

aojea commented:
> error message substring coming from a _specific_ implementation of the container runtime (not from a kubernetes component) seems problematic

at least the implementation is the same for CRIO and Containerd ... I've done it :smile:

brianpursley commented:
> Hrmm... coding kubectl to check for a specific error message substring coming from a _specific_ implementation of the container runtime (not from a kubernetes component) seems problematic... is there any precedent for something like this?

Not really.

But I'm not sure how else to handle this.  The error is unwrapped, so we can't use [IsConnectionReset()](https://github.com/kubernetes/kubernetes/blob/a9593d634c6a053848413e600dadbf974627515f/staging/src/k8s.io/apimachinery/pkg/util/net/util.go#L42-L48) to check for it.

I suppose we could make a new func (in apimachinery?) called `IsProbableConnectionReset(err error)` which checks the string and returns true if it probably is a connection reset error.

This would be like apimachinery's [IsProbableEOF()](https://github.com/kubernetes/kubernetes/blob/e1069c64956a43f628d8ae2fcd9107959747c1c6/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L73-L103), which also checks for specific strings, although at a different network layer for http I know.

WDYT?

brianpursley commented:
> Adding string detection that just _happens_ to match strings emitted by specific CRI instances that _happen_ to be implemented in go and plumbed through multiple layers (CRI → kubelet → apiserver → client) doesn't seem like an appropriate way to detect specific errors. I don't have a great suggestion here that doesn't involve improving the specification of CRI behavior

OK, I understand.  Where would be a good place to start, for improving the CRI spec to provide more structured errors while port forwarding?  It [looks like SIG-NODE manages the CRI](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/container-runtime-interface.md), so should I open an issue and tag SIG-NODE for discussion?

liggitt commented:
yeah, all things CRI go through sig-node

mezhaka commented:
@brianpursley You decided to close? Is there any update from the SIG-NODE direction?

brianpursley commented:
@mezhaka It was on my todo list.  I wanted to investigate more, but I went ahead and [created an issue for sig-node](https://github.com/kubernetes/kubernetes/issues/113568).
